### PR TITLE
doc, test: test and explain service flag handling

### PR DESCRIPTION
**Pull request by mzumsande**
Service flags received from the peer-to-peer network are handled differently, depending on how we receive them.
If received directly from an outbound peer the flags belong to, they replace existing flags.
If received via gossip relay (so that anyone could send them), new flags are added, but existing ones but cannot be overwritten.

Document that and add test coverage for it.

### CODE REVIEW

**Comment by brunoerg** [comment]
Concept ACK

**Review by jonatack — COMMENTED** [review]
ACK d536e5a6325d1885224f36debdcc5245b94efe8a

File: src/addrman.h [hunk 1]
```diff
@@ -111,6 +111,9 @@ class AddrMan
 
     /**
      * Attempt to add one or more addresses to addrman's new table.
+     * If an address already exists in addrman, the existing entry may be updated
+     * (e.g. adding additional service flags). If the existing entry is in the new table,
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/addrman.h:115]: I think that followed from the `@return` doc, but it wasn't so obvious that adding an existing addr to another new bucket also affects the return value; Changed the `@return` description to `true if at least one address is successfully added, or added to an additional bucket. Unaffected by updates.`

File: src/net_processing.cpp [hunk 2]
```diff
@@ -3376,6 +3376,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         vRecv >> CNetAddr::V1(addrMe);
         if (!pfrom.IsInboundConn())
         {
+            // Overwrites potentially existing services. In contrast to this,
+            // unvalidated services received via gossip relay are only
+            // ever added but cannot replace existing ones.
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/net_processing.cpp:3381]: done

File: src/addrman.h [hunk 3]
```diff
@@ -111,6 +111,9 @@ class AddrMan
...
+     * it may be added to more buckets, improving the probability of selection.
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/addrman.h:116]: I think it's better if the header doesn't refer to implementation details. If I remember correctly, this logic used to be in `Add()`, then `Add_()`, then `AddSingle()`, and I don't think it's nice if we have to update `addrman.h` for refactors like that.

File: src/test/addrman_tests.cpp [hunk 4]
```diff
@@ -1095,6 +1095,32 @@ BOOST_AUTO_TEST_CASE(addrman_update_address)
...
+    BOOST_CHECK_EQUAL(vAddr4.size(), 1U);
+    BOOST_CHECK_EQUAL(vAddr4.at(0).nServices, NODE_NETWORK);
+
+    // Promoting to Tried does not affect the service flags
+    BOOST_CHECK(addrman->Good(addr)); // addr has NODE_NONE
+    std::vector<CAddress> vAddr5{addrman->GetAddr(/*max_addresses=*/0, /*max_pct=*/0, /*network=*/std::nullopt)};
+    BOOST_CHECK_EQUAL(vAddr5.size(), 1U);
+    BOOST_CHECK_EQUAL(vAddr5.at(0).nServices, NODE_NETWORK);
+
+    // Adding service flags even works when the addr is in Tried
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/test/addrman_tests.cpp:1119]: similar to above, I don't like referring to implementation details in comments unless really necessary.

**Comment by mzumsande** [comment]
[d536e5a ](https://github.com/bitcoin/bitcoin/commit/d536e5a6325d1885224f36debdcc5245b94efe8a)to [74ebd4d](https://github.com/bitcoin/bitcoin/commit/74ebd4d1359edce82a134dfcd3da9840f8d206e2): Addressed feedback

**Review by furszy — COMMENTED** [review]
ACK 74ebd4d1359edce82a134dfcd3da9840f8d206e2

**Review by brunoerg — APPROVED** [review]
utACK 74ebd4d1359edce82a134dfcd3da9840f8d206e2

**Comment by achow101** [comment]
ACK 74ebd4d1359edce82a134dfcd3da9840f8d206e2

**Pull request merged**